### PR TITLE
[Arm64] Initialize indirection cell for delegates

### LINKED ISSUE #4311: Compiler optimization assumption invalid: FAILED: pMgr != 0
Starting:    System.Reflection.Tests

Assert failure(PID 5960 [0x00001748], Thread: 3988 [0x0f94]): Compiler optimization assumption invalid: FAILED: pMgr != 0

CORECLR! coreclr_shutdown + 0x1DBD0 (0x00007ff8`1dd53750)
CORECLR! GetCLRRuntimeHost + 0x4F0B4C (0x00007ff8`1e31621c)
CORECLR! <no symbol> + 0x0 (0x00007ff8`1dd11a48)
CORECLR! <no symbol> + 0x0 (0x00007ff8`1dd11eb4)
CORECLR! GetCLRRuntimeHost + 0x5C9250 (0x00007ff8`1e3ee920)
CORECLR! GetCLRRuntimeHost + 0xBAA090 (0x00007ff8`1e9cf760)
CORECLR! GetCLRRuntimeHost + 0xB96298 (0x00007ff8`1e9bb968)
CORECLR! GetCLRRuntimeHost + 0xBA626C (0x00007ff8`1e9cb93c)
<no module>! <no symbol> + 0x0 (0x00007ff7`bf92dbc4)
    File: e:\git\coreclr\src\vm\virtualcallstub.cpp Line: 1600
    Image: Z:\corefx\bin\tests\Windows_NT.AnyCPU.Debug\System.Reflection.Tests\dnxcore50\CoreRun.exe

**** MessageBox invoked, title 'CoreRun.exe - Assert Failure (PID 5960, Thread 3988/0x0f94)' ****
  Compiler optimization assumption invalid: FAILED: pMgr != 0

CORECLR! coreclr_shutdown + 0x1DBD0 (0x00007ff8`1dd53750)
CORECLR! GetCLRRuntimeHost + 0x4F0B4C (0x00007ff8`1e31621c)
CORECLR! <no symbol> + 0x0 (0x00007ff8`1dd11a48)
CORECLR! <no symbol> + 0x0 (0x00007ff8`1dd11eb4)
CORECLR! GetCLRRuntimeHost + 0x5C9250 (0x00007ff8`1e3ee920)
CORECLR! GetCLRRuntimeHost + 0xBAA090 (0x00007ff8`1e9cf760)
CORECLR! GetCLRRuntimeHost + 0xB96298 (0x00007ff8`1e9bb968)
CORECLR! GetCLRRuntimeHost + 0xBA626C (0x00007ff8`1e9cb93c)
<no module>! <no symbol> + 0x0 (0x00007ff7`bf92dbc4)

e:\git\coreclr\src\vm\virtualcallstub.cpp, Line: 1600

### PR DESCRIPTION
**Pull request by ramarag**
fixes #4311

@gkhanna79 @rahku PTAL

@dotnet/arm64-contrib FYI

### CODE REVIEW

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by ramarag** [comment]
@dotnet-bot test Windows_NT arm64 Checked

**Comment by rahku** [comment]
This and the previous test that you added for delegate seem to be simple tests for delegate which I am sure were present in legacy test codebase. I am not sure why these were not ported yet or if they were intentionally moved to corefx repo to avoid duplication

File: src/vm/arm64/stubs.cpp [hunk 1]
```diff
@@ -1606,6 +1606,8 @@ VOID StubLinkerCPU::EmitShuffleThunk(ShuffleEntry *pShuffleEntryArray)
     // field and save it in x9. Tailcall to the target method after re-arranging the arguments
     // ldr x9, [x0, #offsetof(DelegateObject, _methodPtrAux)]
     EmitLoadStoreRegImm(eLOAD, IntReg(9), IntReg(0), DelegateObject::GetOffsetOfMethodPtrAux());
+    //add x11, x0, DelegateObject::GetOffsetOfMethodPtrAux() - load the indirection cell into r11
+    EmitAddImm(IntReg(11), IntReg(0), DelegateObject::GetOffsetOfMethodPtrAux());
```
Comment (rahku) [inline src/vm/arm64/stubs.cpp]: I don't see this being done for arm. Does this scenario fail on arm?

**Comment by rahku** [comment]
But I do see the value for these tests to be in coreclr repo as they are not about public surface but runtime features. And I mean that for most of tests for delegates.

File: src/vm/arm64/stubs.cpp [hunk 2]
```diff
@@ -1606,6 +1606,8 @@ VOID StubLinkerCPU::EmitShuffleThunk(ShuffleEntry *pShuffleEntryArray)
     // field and save it in x9. Tailcall to the target method after re-arranging the arguments
     // ldr x9, [x0, #offsetof(DelegateObject, _methodPtrAux)]
     EmitLoadStoreRegImm(eLOAD, IntReg(9), IntReg(0), DelegateObject::GetOffsetOfMethodPtrAux());
+    //add x11, x0, DelegateObject::GetOffsetOfMethodPtrAux() - load the indirection cell into r11
```
Comment (rahku) [inline src/vm/arm64/stubs.cpp]: change r11 to x11. Add comment that VSD resolveworker expects indirection cell to be in x11

File: tests/src/CoreMangLib/system/delegate/VSD/VSD_ResolveWorker.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (rahku) [inline tests/src/CoreMangLib/system/delegate/VSD/VSD_ResolveWorker.cs]: you may clean out the test to remove unncessary stuff. Also maybe openDelegate might be a better name of this test

Comment (rahku) [inline tests/src/CoreMangLib/system/delegate/VSD/VSD_ResolveWorker.cs]: Please open an issue for the test gap for VM side of delegate testing in coreclr repo

Comment (ramarag) [inline tests/src/CoreMangLib/system/delegate/VSD/VSD_ResolveWorker.cs]: will do

**Comment by rahku** [comment]
Otherwise LGTM

File: src/vm/arm64/stubs.cpp (hunk 2 above)
Comment (ramarag) [inline src/vm/arm64/stubs.cpp]: will do

File: src/vm/arm64/stubs.cpp (hunk 1 above)
Comment (ramarag) [inline src/vm/arm64/stubs.cpp]: For Arm this scenario is made to work by erecting a secure delegate frame before calling into ShuffleThunk

**Pull request merged**